### PR TITLE
[5.5] add date and cast check for dirty

### PR DESCRIPTION
**Pull request by golubkovden**
Reopened PR from #18385

add tests

### CODE REVIEW

**Comment by taylorotwell** [comment]
Your tests fail.

**Comment by golubkovden** [comment]
Hm...
Looks like i have same problem with travis as #18417
Will check in local tests.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 1]
```diff
@@ -973,16 +970,35 @@ public function getDirty()
...
+        }
 
-        $original = $this->original[$key];
+        if ($current === $original = $this->getOriginal($key)) {
+            return true;
+        }
+
+        // When check rich this check and current attribute value not equals with original, we should skip next steps
+        // if current is null
+        if (is_null($current)) {
```
Comment (golubkovden) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: This check for nullable may be should move to [fromDateTime](https://github.com/laravel/framework/blob/5.4/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L730) method as like [castAttribute](https://github.com/laravel/framework/blob/5.4/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L458) method.

This hack check looked from [setAttribute](https://github.com/laravel/framework/blob/5.4/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L525)

Comment (daftspunk) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: [Fully agree](https://github.com/octobercms/library/commit/565761d876447b8f2819a0fa1ab48354599eb814#diff-dfa33b774dcd9386d95be910586cc883R354), either that or should to include the original value in the check too:

```
 if (is_null($current) || is_null($original)) {
```

- `is_null($current)` covers updating filled column with NULL
- `is_null($original)` covers updating NULL with filling

Comment (trip-somers) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: The missing `is_null($original)` causes problems for `deleted_at` which is NULL before a model is updated.

**Pull request merged**